app/scripts/controllers/transactions/index.js

```diff
@@ -1084,6 +1150,83 @@ export default class TransactionController extends EventEmitter {
     }
   }
 
+  async confirmExternalTransaction(txMeta, txReceipt, baseFeePerGas) {
+    // add external transaction
+    await this.txStateManager.addExternalTransaction(txMeta);
+
+    const txId = txMeta.id;
+
+    if (!txMeta) {
+      return;
+    }
+
+    try {
+      // It seems that sometimes the numerical values being returned from
+      // this.query.getTransactionReceipt are BN instances and not strings.
+      const gasUsed =
+        typeof txReceipt.gasUsed === 'string'
+          ? txReceipt.gasUsed
+          : txReceipt.gasUsed.toString(16);
+
+      txMeta.txReceipt = {
+        ...txReceipt,
+        gasUsed,
+      };
+
+      if (baseFeePerGas) {
+        txMeta.baseFeePerGas = baseFeePerGas;
+      }
+
+      this.txStateManager.setTxStatusConfirmed(txId);
+      this._markNonceDuplicatesDropped(txId);
+
+      const { submittedTime } = txMeta;
+      const metricsParams = { gas_used: gasUsed };
+
+      if (submittedTime) {
+        metricsParams.completion_time = this._getTransactionCompletionTime(
+          submittedTime,
+        );
+      }
+
+      if (txReceipt.status === '0x0') {
+        metricsParams.status = 'failed on-chain';
```

Comment:
I think the `failed on-chain` error is used in more than one place now, so we may want to make it a constant.